<html><head><meta content="text/html; charset=UTF-8" http-equiv="content-type"><style type="text/css">@import url('https://themes.googleusercontent.com/fonts/css?kit=XGMkxXUZTA64h2imyzu79g');ol.lst-kix_ekhy04fgp052-8.start{counter-reset:lst-ctn-kix_ekhy04fgp052-8 0}.lst-kix_ekhy04fgp052-0>li{counter-increment:lst-ctn-kix_ekhy04fgp052-0}.lst-kix_gup660q9tmf3-7>li:before{content:"-  "}ol.lst-kix_ekhy04fgp052-5.start{counter-reset:lst-ctn-kix_ekhy04fgp052-5 0}.lst-kix_gup660q9tmf3-8>li:before{content:"-  "}.lst-kix_ekhy04fgp052-1>li{counter-increment:lst-ctn-kix_ekhy04fgp052-1}.lst-kix_ddct7fkq5etb-5>li:before{content:"-  "}.lst-kix_qxq34zax48bo-0>li:before{content:"\0025cf  "}.lst-kix_qxq34zax48bo-2>li:before{content:"\0025a0  "}.lst-kix_ddct7fkq5etb-3>li:before{content:"-  "}.lst-kix_ddct7fkq5etb-7>li:before{content:"-  "}.lst-kix_ddct7fkq5etb-2>li:before{content:"-  "}.lst-kix_ddct7fkq5etb-6>li:before{content:"-  "}.lst-kix_qxq34zax48bo-3>li:before{content:"\0025cf  "}.lst-kix_397xt09pl4qx-6>li:before{content:"-  "}.lst-kix_397xt09pl4qx-5>li:before{content:"-  "}.lst-kix_ddct7fkq5etb-4>li:before{content:"-  "}.lst-kix_qxq34zax48bo-1>li:before{content:"\0025cb  "}.lst-kix_397xt09pl4qx-3>li:before{content:"-  "}.lst-kix_397xt09pl4qx-2>li:before{content:"-  "}.lst-kix_397xt09pl4qx-4>li:before{content:"-  "}.lst-kix_ekhy04fgp052-3>li{counter-increment:lst-ctn-kix_ekhy04fgp052-3}.lst-kix_qxq34zax48bo-8>li:before{content:"\0025a0  "}.lst-kix_qxq34zax48bo-7>li:before{content:"\0025cb  "}.lst-kix_ddct7fkq5etb-1>li:before{content:"-  "}.lst-kix_397xt09pl4qx-0>li:before{content:"-  "}.lst-kix_qxq34zax48bo-4>li:before{content:"\0025cb  "}.lst-kix_qxq34zax48bo-6>li:before{content:"\0025cf  "}.lst-kix_397xt09pl4qx-1>li:before{content:"-  "}.lst-kix_ddct7fkq5etb-0>li:before{content:"-  "}.lst-kix_qxq34zax48bo-5>li:before{content:"\0025a0  "}ul.lst-kix_m6ci32kd50uv-8{list-style-type:none}ul.lst-kix_m6ci32kd50uv-7{list-style-type:none}ul.lst-kix_m6ci32kd50uv-6{list-style-type:none}ul.lst-kix_m6ci32kd50uv-1{list-style-type:none}.lst-kix_m6ci32kd50uv-0>li:before{content:"-  "}ul.lst-kix_m6ci32kd50uv-0{list-style-type:none}.lst-kix_m6ci32kd50uv-1>li:before{content:"-  "}ul.lst-kix_gup660q9tmf3-0{list-style-type:none}ul.lst-kix_m6ci32kd50uv-5{list-style-type:none}ul.lst-kix_gup660q9tmf3-1{list-style-type:none}ul.lst-kix_m6ci32kd50uv-4{list-style-type:none}ul.lst-kix_gup660q9tmf3-2{list-style-type:none}ul.lst-kix_m6ci32kd50uv-3{list-style-type:none}ul.lst-kix_gup660q9tmf3-3{list-style-type:none}ul.lst-kix_m6ci32kd50uv-2{list-style-type:none}ul.lst-kix_gup660q9tmf3-4{list-style-type:none}ul.lst-kix_yng2as9iogot-4{list-style-type:none}ul.lst-kix_pbqqzrwh3zqo-0{list-style-type:none}ul.lst-kix_yng2as9iogot-5{list-style-type:none}ul.lst-kix_yng2as9iogot-2{list-style-type:none}ul.lst-kix_yng2as9iogot-3{list-style-type:none}ul.lst-kix_yng2as9iogot-8{list-style-type:none}ul.lst-kix_yng2as9iogot-6{list-style-type:none}ul.lst-kix_yng2as9iogot-7{list-style-type:none}.lst-kix_397xt09pl4qx-7>li:before{content:"-  "}.lst-kix_ekhy04fgp052-4>li{counter-increment:lst-ctn-kix_ekhy04fgp052-4}.lst-kix_397xt09pl4qx-8>li:before{content:"-  "}ol.lst-kix_ekhy04fgp052-0.start{counter-reset:lst-ctn-kix_ekhy04fgp052-0 0}ul.lst-kix_yng2as9iogot-0{list-style-type:none}ul.lst-kix_yng2as9iogot-1{list-style-type:none}ol.lst-kix_ekhy04fgp052-6.start{counter-reset:lst-ctn-kix_ekhy04fgp052-6 0}.lst-kix_ekhy04fgp052-7>li{counter-increment:lst-ctn-kix_ekhy04fgp052-7}ul.lst-kix_6thhdslfbwlj-7{list-style-type:none}ul.lst-kix_6thhdslfbwlj-8{list-style-type:none}ul.lst-kix_397xt09pl4qx-8{list-style-type:none}.lst-kix_ekhy04fgp052-1>li:before{content:"" counter(lst-ctn-kix_ekhy04fgp052-1,lower-latin) ") "}.lst-kix_ekhy04fgp052-3>li:before{content:"(" counter(lst-ctn-kix_ekhy04fgp052-3,decimal) ") "}.lst-kix_ekhy04fgp052-2>li:before{content:"" counter(lst-ctn-kix_ekhy04fgp052-2,lower-roman) ") "}.lst-kix_pbqqzrwh3zqo-0>li:before{content:"\0025cf  "}ul.lst-kix_pbqqzrwh3zqo-7{list-style-type:none}ul.lst-kix_pbqqzrwh3zqo-8{list-style-type:none}.lst-kix_m6ci32kd50uv-8>li:before{content:"-  "}ol.lst-kix_ekhy04fgp052-7.start{counter-reset:lst-ctn-kix_ekhy04fgp052-7 0}.lst-kix_pbqqzrwh3zqo-1>li:before{content:"\0025cb  "}ul.lst-kix_pbqqzrwh3zqo-5{list-style-type:none}ul.lst-kix_pbqqzrwh3zqo-6{list-style-type:none}.lst-kix_pbqqzrwh3zqo-2>li:before{content:"\0025a0  "}ul.lst-kix_pbqqzrwh3zqo-3{list-style-type:none}.lst-kix_ekhy04fgp052-7>li:before{content:"" counter(lst-ctn-kix_ekhy04fgp052-7,lower-latin) ". "}ul.lst-kix_pbqqzrwh3zqo-4{list-style-type:none}ul.lst-kix_pbqqzrwh3zqo-1{list-style-type:none}.lst-kix_ekhy04fgp052-8>li:before{content:"" counter(lst-ctn-kix_ekhy04fgp052-8,lower-roman) ". "}ul.lst-kix_pbqqzrwh3zqo-2{list-style-type:none}.lst-kix_ekhy04fgp052-0>li:before{content:"" counter(lst-ctn-kix_ekhy04fgp052-0,upper-latin) ") "}ul.lst-kix_gup660q9tmf3-5{list-style-type:none}ol.lst-kix_ekhy04fgp052-5{list-style-type:none}.lst-kix_m6ci32kd50uv-4>li:before{content:"-  "}.lst-kix_m6ci32kd50uv-5>li:before{content:"-  "}ul.lst-kix_gup660q9tmf3-6{list-style-type:none}ol.lst-kix_ekhy04fgp052-4{list-style-type:none}ul.lst-kix_qxq34zax48bo-8{list-style-type:none}ul.lst-kix_gup660q9tmf3-7{list-style-type:none}ol.lst-kix_ekhy04fgp052-3{list-style-type:none}ul.lst-kix_qxq34zax48bo-7{list-style-type:none}ul.lst-kix_gup660q9tmf3-8{list-style-type:none}ol.lst-kix_ekhy04fgp052-2{list-style-type:none}ul.lst-kix_qxq34zax48bo-6{list-style-type:none}ol.lst-kix_ekhy04fgp052-1{list-style-type:none}.lst-kix_ekhy04fgp052-6>li{counter-increment:lst-ctn-kix_ekhy04fgp052-6}ul.lst-kix_qxq34zax48bo-5{list-style-type:none}.lst-kix_m6ci32kd50uv-2>li:before{content:"-  "}.lst-kix_m6ci32kd50uv-3>li:before{content:"-  "}.lst-kix_m6ci32kd50uv-6>li:before{content:"-  "}.lst-kix_m6ci32kd50uv-7>li:before{content:"-  "}ol.lst-kix_ekhy04fgp052-0{list-style-type:none}ul.lst-kix_qxq34zax48bo-4{list-style-type:none}.lst-kix_ekhy04fgp052-6>li:before{content:"" counter(lst-ctn-kix_ekhy04fgp052-6,decimal) ". "}ul.lst-kix_qxq34zax48bo-3{list-style-type:none}ul.lst-kix_qxq34zax48bo-2{list-style-type:none}ul.lst-kix_397xt09pl4qx-0{list-style-type:none}ol.lst-kix_ekhy04fgp052-1.start{counter-reset:lst-ctn-kix_ekhy04fgp052-1 0}ol.lst-kix_ekhy04fgp052-4.start{counter-reset:lst-ctn-kix_ekhy04fgp052-4 0}.lst-kix_ekhy04fgp052-5>li:before{content:"(" counter(lst-ctn-kix_ekhy04fgp052-5,lower-roman) ") "}ul.lst-kix_qxq34zax48bo-1{list-style-type:none}ul.lst-kix_397xt09pl4qx-1{list-style-type:none}ul.lst-kix_qxq34zax48bo-0{list-style-type:none}ul.lst-kix_397xt09pl4qx-2{list-style-type:none}ul.lst-kix_397xt09pl4qx-3{list-style-type:none}ul.lst-kix_397xt09pl4qx-4{list-style-type:none}ul.lst-kix_397xt09pl4qx-5{list-style-type:none}ol.lst-kix_ekhy04fgp052-8{list-style-type:none}ul.lst-kix_397xt09pl4qx-6{list-style-type:none}ol.lst-kix_ekhy04fgp052-7{list-style-type:none}ul.lst-kix_397xt09pl4qx-7{list-style-type:none}.lst-kix_ekhy04fgp052-4>li:before{content:"(" counter(lst-ctn-kix_ekhy04fgp052-4,lower-latin) ") "}ol.lst-kix_ekhy04fgp052-6{list-style-type:none}ol.lst-kix_ekhy04fgp052-3.start{counter-reset:lst-ctn-kix_ekhy04fgp052-3 0}.lst-kix_6thhdslfbwlj-0>li:before{content:"-  "}.lst-kix_yng2as9iogot-8>li:before{content:"-  "}.lst-kix_pbqqzrwh3zqo-8>li:before{content:"\0025a0  "}.lst-kix_pbqqzrwh3zqo-4>li:before{content:"\0025cb  "}.lst-kix_6thhdslfbwlj-5>li:before{content:"-  "}.lst-kix_pbqqzrwh3zqo-3>li:before{content:"\0025cf  "}.lst-kix_pbqqzrwh3zqo-5>li:before{content:"\0025a0  "}.lst-kix_yng2as9iogot-3>li:before{content:"-  "}.lst-kix_6thhdslfbwlj-3>li:before{content:"-  "}.lst-kix_6thhdslfbwlj-4>li:before{content:"-  "}.lst-kix_yng2as9iogot-4>li:before{content:"-  "}.lst-kix_yng2as9iogot-7>li:before{content:"-  "}.lst-kix_6thhdslfbwlj-1>li:before{content:"-  "}.lst-kix_6thhdslfbwlj-2>li:before{content:"-  "}.lst-kix_pbqqzrwh3zqo-7>li:before{content:"\0025cb  "}.lst-kix_pbqqzrwh3zqo-6>li:before{content:"\0025cf  "}.lst-kix_yng2as9iogot-5>li:before{content:"-  "}.lst-kix_yng2as9iogot-6>li:before{content:"-  "}.lst-kix_ddct7fkq5etb-8>li:before{content:"-  "}ul.lst-kix_6thhdslfbwlj-5{list-style-type:none}ul.lst-kix_ddct7fkq5etb-2{list-style-type:none}ul.lst-kix_6thhdslfbwlj-6{list-style-type:none}ul.lst-kix_ddct7fkq5etb-1{list-style-type:none}ul.lst-kix_6thhdslfbwlj-3{list-style-type:none}.lst-kix_ekhy04fgp052-5>li{counter-increment:lst-ctn-kix_ekhy04fgp052-5}ul.lst-kix_ddct7fkq5etb-4{list-style-type:none}ul.lst-kix_6thhdslfbwlj-4{list-style-type:none}.lst-kix_gup660q9tmf3-1>li:before{content:"-  "}ul.lst-kix_ddct7fkq5etb-3{list-style-type:none}ul.lst-kix_6thhdslfbwlj-1{list-style-type:none}.lst-kix_ekhy04fgp052-2>li{counter-increment:lst-ctn-kix_ekhy04fgp052-2}ul.lst-kix_6thhdslfbwlj-2{list-style-type:none}ul.lst-kix_ddct7fkq5etb-0{list-style-type:none}ul.lst-kix_6thhdslfbwlj-0{list-style-type:none}.lst-kix_gup660q9tmf3-0>li:before{content:"-  "}.lst-kix_gup660q9tmf3-4>li:before{content:"-  "}.lst-kix_gup660q9tmf3-5>li:before{content:"-  "}.lst-kix_yng2as9iogot-1>li:before{content:"-  "}ul.lst-kix_ddct7fkq5etb-6{list-style-type:none}ol.lst-kix_ekhy04fgp052-2.start{counter-reset:lst-ctn-kix_ekhy04fgp052-2 0}ul.lst-kix_ddct7fkq5etb-5{list-style-type:none}li.li-bullet-0:before{margin-left:-18pt;white-space:nowrap;display:inline-block;min-width:18pt}ul.lst-kix_ddct7fkq5etb-8{list-style-type:none}.lst-kix_gup660q9tmf3-6>li:before{content:"-  "}.lst-kix_yng2as9iogot-2>li:before{content:"-  "}.lst-kix_ekhy04fgp052-8>li{counter-increment:lst-ctn-kix_ekhy04fgp052-8}ul.lst-kix_ddct7fkq5etb-7{list-style-type:none}.lst-kix_6thhdslfbwlj-6>li:before{content:"-  "}.lst-kix_6thhdslfbwlj-7>li:before{content:"-  "}.lst-kix_6thhdslfbwlj-8>li:before{content:"-  "}.lst-kix_yng2as9iogot-0>li:before{content:"-  "}.lst-kix_gup660q9tmf3-3>li:before{content:"-  "}.lst-kix_gup660q9tmf3-2>li:before{content:"-  "}ol{margin:0;padding:0}table td,table th{padding:0}.c9{margin-left:18pt;padding-top:3pt;padding-bottom:0pt;line-height:1.0;orphans:2;widows:2;text-align:left}.c23{color:#434343;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c18{color:#434343;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:14pt;font-family:"Arial";font-style:normal}.c31{color:#ce9178;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:9pt;font-family:"Courier New";font-style:normal}.c25{color:#999999;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c21{margin-left:36pt;padding-top:3pt;padding-bottom:0pt;line-height:1.0;orphans:2;widows:2;text-align:left}.c27{color:#1d1c1d;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c10{padding-top:20pt;padding-bottom:6pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.c28{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:26pt;font-family:"Arial";font-style:normal}.c30{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:14pt;font-family:"Arial";font-style:normal}.c34{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:8pt;font-family:"Arial";font-style:normal}.c4{color:#000000;font-weight:700;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c3{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:16pt;font-family:"Arial";font-style:normal}.c2{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:8pt;font-family:"Roboto Mono";font-style:normal}.c33{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:20pt;font-family:"Arial";font-style:normal}.c19{padding-top:18pt;padding-bottom:6pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.c8{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:10.5pt;font-family:"Arial";font-style:normal}.c32{margin-left:18pt;padding-top:3pt;padding-bottom:4pt;line-height:1.0;orphans:2;widows:2;text-align:left}.c0{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c24{padding-top:0pt;padding-bottom:3pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.c1{padding-top:16pt;padding-bottom:4pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.c20{padding-top:10pt;padding-bottom:0pt;line-height:1.0;orphans:2;widows:2;text-align:left}.c39{padding-top:4pt;padding-bottom:0pt;line-height:1.0;orphans:2;widows:2;text-align:left}.c6{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c41{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:right}.c38{padding-top:12pt;padding-bottom:12pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c13{color:#000000;text-decoration:none;vertical-align:baseline;font-size:11pt;font-style:normal}.c26{text-decoration-skip-ink:none;-webkit-text-decoration-skip:none;color:#1155cc;text-decoration:underline}.c42{text-decoration-skip-ink:none;-webkit-text-decoration-skip:none;color:#0000ff;text-decoration:underline}.c17{font-size:10.5pt;font-family:"Courier New";font-weight:400}.c37{background-color:#ffffff;max-width:452.6pt;padding:72pt 72pt 72pt 70.9pt}.c12{font-weight:400;font-family:"Roboto Mono"}.c14{margin-left:36pt;padding-left:0pt}.c11{padding:0;margin:0}.c5{color:inherit;text-decoration:inherit}.c16{margin-left:72pt;padding-left:0pt}.c40{background-color:#ffffff}.c22{font-size:10.5pt}.c43{font-size:8pt}.c7{height:11pt}.c29{margin-left:36pt}.c36{color:#1d1c1d}.c35{text-indent:36pt}.c15{font-weight:700}.title{padding-top:0pt;color:#000000;font-size:26pt;padding-bottom:3pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.subtitle{padding-top:0pt;color:#666666;font-size:15pt;padding-bottom:16pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}li{color:#000000;font-size:11pt;font-family:"Arial"}p{margin:0;color:#000000;font-size:11pt;font-family:"Arial"}h1{padding-top:20pt;color:#000000;font-size:20pt;padding-bottom:6pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h2{padding-top:18pt;color:#000000;font-size:16pt;padding-bottom:6pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h3{padding-top:16pt;color:#434343;font-size:14pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h4{padding-top:14pt;color:#666666;font-size:12pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h5{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h6{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;font-style:italic;orphans:2;widows:2;text-align:left}</style></head><body class="c37 doc-content"><div><p class="c6"><span class="c25">Hunt Lab Small RNA Pipeline Docs&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Version 1.0</span></p><hr><p class="c6 c7"><span class="c25"></span></p><p class="c6 c7"><span class="c25"></span></p></div><p class="c24 title" id="h.npjp1uz2wdtu"><span class="c28">Hunt Lab Small RNA Pipeline Docs</span></p><p class="c6"><span class="c0">Kieran Reynolds</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">This tool is designed to automate a lot of the common bioinformatics performed on small RNA by the Hunt Lab. It is deliberately designed to be highly configurable so it can be easily applied to a wide range of scenarios.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c39"><span class="c4"><a class="c5" href="#h.5qmj8kdxqedp">Installation</a></span><span class="c4">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c4"><a class="c5" href="#h.5qmj8kdxqedp">2</a></span></p><p class="c20"><span class="c4"><a class="c5" href="#h.c4hm04k9vpyc">Tasks</a></span><span class="c4">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c4"><a class="c5" href="#h.c4hm04k9vpyc">2</a></span></p><p class="c20"><span class="c4"><a class="c5" href="#h.5qeusqizjjyg">Example Usage</a></span><span class="c4">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c4"><a class="c5" href="#h.5qeusqizjjyg">3</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.dvt908bt3b0l">Process</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.dvt908bt3b0l">3</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.q7r4a262np6p">Sort</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.q7r4a262np6p">4</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.87icg528lpfn">Unitas</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.87icg528lpfn">5</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.wwf4w96dtbfu">Extract Non-Coding</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.wwf4w96dtbfu">6</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.1fny8pcaxh8z">Target Identification</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.1fny8pcaxh8z">6</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.e77dx6e6zu8e">All</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.e77dx6e6zu8e">7</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.g9ms4p1r7z4p">Plotting Genome Location</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.g9ms4p1r7z4p">7</a></span></p><p class="c21"><span class="c0"><a class="c5" href="#h.v8b0nssch3a2">Basic Plot</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.v8b0nssch3a2">7</a></span></p><p class="c21"><span class="c0"><a class="c5" href="#h.3ax5siuj26a8">Dealing with Multiple Scaffolds</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.3ax5siuj26a8">9</a></span></p><p class="c20"><span class="c4"><a class="c5" href="#h.m9e2tyi06fss">Further Processing Options</a></span><span class="c4">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c4"><a class="c5" href="#h.m9e2tyi06fss">11</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.mu0rcgmaopdr">Signature Identification</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.mu0rcgmaopdr">11</a></span></p><p class="c21"><span class="c0"><a class="c5" href="#h.j40p9cyavwlb">AT/GC Richness</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.j40p9cyavwlb">11</a></span></p><p class="c21"><span class="c0"><a class="c5" href="#h.4r7fhoddcf9x">Upstream motif (e.g. piRNA)</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.4r7fhoddcf9x">11</a></span></p><p class="c21"><span class="c0"><a class="c5" href="#h.uqa0lrhum6ba">Ping Pong signature</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.uqa0lrhum6ba">11</a></span></p><p class="c21"><span class="c0"><a class="c5" href="#h.k4lth3h5hxsf">Dicer Signature (StepRNA)</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.k4lth3h5hxsf">12</a></span></p><p class="c21"><span class="c0"><a class="c5" href="#h.wcf5w2j3zywp">Same Strand Overlap</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.wcf5w2j3zywp">13</a></span></p><p class="c21"><span class="c0"><a class="c5" href="#h.po793ufjxkq3">miRNA</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.po793ufjxkq3">13</a></span></p><p class="c20"><span class="c4"><a class="c5" href="#h.ckvkzxv8xboy">Notes on Settings</a></span><span class="c4">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c4"><a class="c5" href="#h.ckvkzxv8xboy">13</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.dqg4gw5x0t24">Configuration file</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.dqg4gw5x0t24">13</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.8xpup46t5df5">Manually setting parameters of internal commands</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.8xpup46t5df5">14</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.u09whqqmb0i9">Suppressing excess output</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.u09whqqmb0i9">14</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.c34jicjkp0we">Troubleshooting</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.c34jicjkp0we">15</a></span></p><p class="c20"><span class="c15"><a class="c5" href="#h.yn1evvqauc8s">CLI Reference</a></span><span class="c15">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c15"><a class="c5" href="#h.yn1evvqauc8s">16</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.rt1yuq1t17b5">Process</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.rt1yuq1t17b5">16</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.v1awwydjuyb5">Sort</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.v1awwydjuyb5">17</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.4iigvduhoact">ExtractNC</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.4iigvduhoact">18</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.9bjdtm5xm2gh">Unitas</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.9bjdtm5xm2gh">18</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.h79krq96jgla">TargetID</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.h79krq96jgla">19</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.jyq65jwuj5bd">Label for Unitas</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.jyq65jwuj5bd">20</a></span></p><p class="c9"><span class="c0"><a class="c5" href="#h.h5pmlelo7os6">Build Coord Files</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.h5pmlelo7os6">20</a></span></p><p class="c32"><span class="c0"><a class="c5" href="#h.mkenrz55dxfs">Overlap SS</a></span><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c0"><a class="c5" href="#h.mkenrz55dxfs">21</a></span></p><h1 class="c10" id="h.5qmj8kdxqedp"><span class="c33">Installation</span></h1><p class="c6"><span>You can obtain the code from </span><span class="c26"><a class="c5" href="https://www.google.com/url?q=https://github.com/kieranr51/HuntLab-smallRNA&amp;sa=D&amp;source=editors&amp;ust=1664380231696777&amp;usg=AOvVaw3NMvjGhFLvqjN5KpROvIRc">https://github.com/kieranr51/HuntLab-smallRNA</a></span><span class="c0">. To download, click the code button and copy the URL. You can then download it to a computer with git installed on it by opening the command line and running:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ git clone &lt;copied_url&gt;</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>It is recommended to use conda to install this pipeline. To perform the installation, ensure conda is installed on the machine, navigate to the directory containing the code and </span><span class="c12">environment.yml</span><span class="c0">&nbsp;file, then run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ conda env create -f environment.yml</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>This will create a conda environment called </span><span class="c12">huntlab-smallrna</span><span class="c0">. This can then be activated each time you want to use the software with the command:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ conda activate huntlab-smallrna</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>By default this provides the </span><span class="c12">hlsmallrna</span><span>&nbsp;executable. This provides all of the tasks listed below, each one can be run by typing </span><span class="c12">hlsmallrna</span><span>&nbsp;followed by the name and options for the task.</span></p><h1 class="c10" id="h.c4hm04k9vpyc"><span class="c33">Tasks</span></h1><p class="c6"><span class="c0">Currently the following tasks have been implemented in it:<br></span></p><ul class="c11 lst-kix_yng2as9iogot-0 start"><li class="c6 c14 li-bullet-0"><span class="c15">Process</span><span class="c0">&nbsp;- Trim adapters off raw RNA-Seq data, run a quality control report and trim of any bases that fall below a specified threshold from all the RNA</span></li></ul><p class="c6 c7 c29"><span class="c0"></span></p><ul class="c11 lst-kix_yng2as9iogot-0"><li class="c6 c14 li-bullet-0"><span class="c15">Sort</span><span class="c0">&nbsp;- Align the RNA against a given reference, removing any sequences that fail to align, then split them into files based on sequence length to allow for easy further processing. Can also remove sequences below a minimum length and above a maximum length if desired</span></li></ul><p class="c6 c7 c29"><span class="c0"></span></p><ul class="c11 lst-kix_yng2as9iogot-0"><li class="c6 c14 li-bullet-0"><span class="c15">ExtractNC</span><span class="c0">&nbsp;- Using a genome and GFF file containing annotations, extract the region that is transcribed, but not a coding region</span></li></ul><p class="c6 c7 c29"><span class="c0"></span></p><ul class="c11 lst-kix_yng2as9iogot-0"><li class="c6 c14 li-bullet-0"><span class="c15">Unitas</span><span class="c0">&nbsp;- Run a directory of RNA files (e.g. output of sort step) though unitas to classify the types of RNA, then combine the summaries into one spreadsheet for easy viewing and manipulation</span></li></ul><p class="c6 c7 c29"><span class="c0"></span></p><ul class="c11 lst-kix_yng2as9iogot-0"><li class="c6 c14 li-bullet-0"><span class="c15">TargetID</span><span class="c0">&nbsp;- Identify which of a set of potential targets a set of small RNA are targeting by reverse complimenting them and aligning them to the potential targets</span></li></ul><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">More in depth documentation on each of these steps, starting with examples and suggestions for further processing. Further down there is a CLI Reference, that contains general help for each command, including what files are required and what outputs you get from each step.</span></p><h1 class="c10" id="h.5qeusqizjjyg"><span class="c33">Example Usage</span></h1><p class="c6"><span>To use this pipeline, you will need at least a FASTQ file containing raw small RNA-Seq reads. Then depending on what you want to achieve, you can run different steps of the pipeline. If you ever get stuck and need to see the options of a particular stage, all parts have their own help message, that can be seen by passing </span><span class="c12">--help</span><span class="c0">&nbsp;after the name of the stage. </span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>Before starting though, you will need to create a configuration file that allows you to specify some parameters in advance, so you don&rsquo;t have to keep retyping them. By default, it looks for these in a file in your current directory called </span><span class="c12">config.toml, </span><span>though alternatives can be specified by passing the name to the </span><span class="c12">-C </span><span>argument before the name of the stage to run. More details on this can be found in the &lsquo;Configuration file&rsquo; section below, but for now setup a simple file to specify an output directory of </span><span class="c12">output/</span><span>&nbsp;by putting the following in </span><span class="c13 c12">config.toml:</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span class="c13 c12">[general]</span></p><p class="c6"><span class="c12">output_directory = </span><span class="c13 c12">&quot;./output&quot;</span></p><p class="c6 c7"><span class="c31"></span></p><p class="c6"><span>In addition, check that all of the commands can be run by just typing their name in the command line, as this is how the program tries to use them by default. If they can&rsquo;t, you can set a custom path by setting the relevant </span><span class="c12">path_to_</span><span>&nbsp;variable in the configuration file. For example if unitas is run with </span><span class="c12">unitas.pl</span><span>&nbsp;instead of just </span><span class="c12">unitas</span><span class="c0">&nbsp;add the following to the bottom of the config file:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">[cli-tools]</span></p><p class="c6"><span class="c13 c12">[cli-tools.unitas]</span></p><p class="c6"><span class="c12">path_to_unitas = &quot;unitas.pl&quot;</span></p><h2 class="c19" id="h.dvt908bt3b0l"><span class="c3">Process</span></h2><p class="c6"><span class="c0">The first stage is Process. This is for if you have raw data that hasn&rsquo;t had it&rsquo;s adapters trimmed and been quality filtered, so if your data has had this done to it already, this step can be skipped. To run you need a FASTQ file containing raw small RNA and the sequence and end of any adapters attached during sequencing.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">Once you have this data, to include the trimming in the run, you need to set one of three flags with the adapter sequence, depending on the position of the particular adapter. </span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>If your adapter is on the 3&rsquo; end only, use </span><span class="c12">-a; </span><span>if it is from the </span><span>&nbsp;5&rsquo; end only, use </span><span class="c12">-g </span><span>and if it could be on either end use </span><span class="c12">-b. </span><span>If you don&rsquo;t need adapters trimmed, just don&rsquo;t specify any of them and the step will be skipped. For example, if your small RNAs are in </span><span class="c12">smallRNA.fastq </span><span>and you want to trim the adapter sequence </span><span class="c12">AGCATA</span><span class="c0">&nbsp;from the 3&rsquo; end only, you would run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ hlsmallrna process -a AGCATA smallRNA.fastq</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>In addition, this step runs FastQC to produce a report on the quality of the small RNA and automatically cuts anything with a lower quartile of quality below a cutoff. By default, this is set to 20, but can be changed by the user using the </span><span class="c12">-c </span><span>flag. If you do not want this, you can set </span><span class="c12">-c 0 </span><span class="c0">and the FastQC step will be skipped for efficiency. For example if you want to change the cutoff for the last command to 5, you could run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ hlsmallrna process -c 5 -a AGCATA smallRNA.fastq</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>This produces a number of files in the output directory, importantly </span><span class="c12">cut_sequences.fastq </span><span class="c0">has the sequences that have adapters trimmed and low quality parts removed. (See CLI Reference section for information on other output for all of the commands)</span></p><h2 class="c19" id="h.q7r4a262np6p"><span class="c3">Sort</span></h2><p class="c6"><span class="c0">The second stage is Sort. This stage aligns the small RNAs to the genome of the organism and filters out any that don&rsquo;t align, to remove contamination. Then it splits the remaining reads into FASTQ files by length, to a-x llow for easier classification. </span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>To run this, you need a FASTQ file of processed small RNA and a FASTA file containing the cdgenome of the organism of interest. For example, if your small RNA are in </span><span class="c12">smallRNA.fastq</span><span>&nbsp;and your genome is in </span><span class="c12">genome.fasta </span><span class="c0">you can run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ hlsmallrna sort smallRNA.fastq genome.fasta</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>If you are only interested in a subset of lengths of small RNA, this step can be set to restrict the lengths it outputs with </span><span class="c12">-l </span><span>for minimum and </span><span class="c12">-x </span><span class="c0">for maximum. For example, if you only want small RNA with lengths between 12 and 30, you could run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ hlsmallrna sort -l 12 -x 30 smallRNA.fastq genome.fasta</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span class="c0">This produces a few useful files in the output directory: </span></p><p class="c6 c7"><span class="c0"></span></p><ul class="c11 lst-kix_gup660q9tmf3-0 start"><li class="c6 c14 li-bullet-0"><span class="c12">binned_rna/ - </span><span class="c0">directory containing one fastq file for each length of small RNA</span></li></ul><p class="c6 c7 c29"><span class="c0"></span></p><ul class="c11 lst-kix_gup660q9tmf3-0"><li class="c6 c14 li-bullet-0"><span class="c12 c22">rna_length_report.csv</span><span class="c17">&nbsp;- </span><span class="c8">table showing a summary of the RNAs by length and first base</span></li></ul><p class="c6 c7 c29"><span class="c8"></span></p><ul class="c11 lst-kix_gup660q9tmf3-0"><li class="c6 c14 li-bullet-0"><span class="c12 c22">baseplot.png - </span><span class="c22">plot of the length and first base of the RNAs</span></li></ul><h2 class="c19" id="h.87icg528lpfn"><span class="c3">Unitas</span></h2><p class="c6"><span class="c0">The third stage is Unitas. This runs unitas on a directory containing a range of lengths of small RNA to classify the origin of the RNA. To do this, you will need files containing reference sets of the genome targets you wish to find, with labels in their IDs and a directory containing small RNA of different lengths, similar to the output of sort.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>You can set the reference file using the </span><span class="c12">-r</span><span>&nbsp;command line argument, but I think it is easier to set them in the configuration file, as it makes the command shorter to type. If your reference files are in the same directory you are running in and called </span><span class="c12">ref1.fasta, ref2.fasta </span><span>and </span><span class="c12">ref3.fasta</span><span class="c0">, you can add the following to the config file:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">[command]</span></p><p class="c6"><span class="c13 c12">refseq = [</span></p><p class="c6"><span class="c13 c12">&nbsp; &nbsp; &quot;ref1.fasta&quot;,</span></p><p class="c6"><span class="c13 c12">&nbsp; &nbsp; &quot;ref2.fasta&quot;,</span></p><p class="c6"><span class="c13 c12">&nbsp; &nbsp; &quot;ref3.fasta&quot;</span></p><p class="c6"><span class="c13 c12">]</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>Note that all of these reference sequences should be labelled with the name of the category they belong to and the | (bar or pipe character) before the start of their ID. If this hasn&rsquo;t been done and all of the sequences in a FASTA file should be labelled the same, you can use the </span><span class="c12">label_for_unitas </span><span>script that was installed with the small RNA pipeline. For example if you want to label the sequences in </span><span class="c12">unlabelledTEs.fasta </span><span>with the label TE, with the result produced in </span><span class="c12">teRef.fasta</span><span class="c0">, you could run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ label_for_unitas &quot;TE&quot; unlabelledTEs.fasta -o teRef.fasta</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>Once the correct reference file are set, if your small RNA, split by length, are in </span><span class="c12">output/binned_rna, </span><span class="c0">you can run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ &nbsp;hlsmallrna unitas output/binned_rna</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">This produces a useful summary and graph in the output directory:</span></p><p class="c6 c7"><span class="c0"></span></p><ul class="c11 lst-kix_397xt09pl4qx-0 start"><li class="c6 c14 li-bullet-0"><span class="c12">unitasGraph.png - </span><span class="c0">graph showing the categories allocated by unitas against the length of the small RNA</span></li></ul><p class="c6 c7 c29"><span class="c0"></span></p><ul class="c11 lst-kix_397xt09pl4qx-0"><li class="c6 c14 li-bullet-0"><span class="c12">unitas_graph_data.csv - </span><span class="c0">summary of the raw numbers produced from the unitas runs, used to produce the graph</span></li></ul><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">If you want to get small RNA that are derived from genes, you will need to obtain both of the mRNA and CDS regions and pass them to speciled arguments in the pipeline so they can be combined. This can either be done using the config file as followers or with the corresponding command line arguments (--cds and --unspliced-transcriptome):</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">[command]</span></p><p class="c6"><span class="c13 c12">refseq = [</span></p><p class="c6"><span class="c13 c12">&nbsp; &nbsp; &quot;ref1.fasta&quot;,</span></p><p class="c6"><span class="c13 c12">&nbsp; &nbsp; &quot;ref2.fasta&quot;,</span></p><p class="c6"><span class="c13 c12">&nbsp; &nbsp; &quot;ref3.fasta&quot;</span></p><p class="c6"><span class="c13 c12">]</span></p><p class="c6"><span class="c13 c12">cds = &ldquo;cds_sequences.fasta&rdquo;</span></p><p class="c6"><span class="c13 c12">unspliced_transcriptome = &ldquo;unspliced_transcriptome.fasta&rdquo;</span></p><p class="c6 c7"><span class="c0"></span></p><h2 class="c19" id="h.wwf4w96dtbfu"><span class="c3">Extract Non-Coding</span></h2><p class="c6"><span>Sometimes you want to see if a small RNA comes from a non-coding region of the transcriptome, so the pipeline contains a utility command to automatically extract this using a genome and some annotations. At minimum, these annotations need to contain a transcribed region labelled with &lsquo;mRNA&rsquo; and a coding region within the transcribed region labelled &lsquo;CDS&rsquo;. Most of the ones on Wormbase P</span><span>arasite</span><span class="c0">&nbsp;are fine for this purpose.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>As an example, to extract the regions from </span><span class="c12">genome.fasta </span><span>using the annotations in </span><span class="c12">annotations.gff </span><span class="c0">you could run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ hlsmallrna extractnc genome.fasta annotations.gff</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>This will produce a FASTA file containing the non-coding regions, complete with unitas labels applied, called </span><span class="c12">noncoding.fasta </span><span>in the output directory.</span></p><h2 class="c19" id="h.1fny8pcaxh8z"><span class="c3">Target Identification</span></h2><p class="c6"><span class="c0">The fourth and final main step is Target Identification. This takes a set of small RNA and a set of potential targets and produces a list of genome features that are targeted by each small RNA. </span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>As with unitas, you can set target files using the command line option </span><span class="c12">-t </span><span>but it is easier to add them to the configuration file. For example if your target files are </span><span class="c12">target1.fasta, target2.fasta </span><span>and </span><span class="c12">target3.fasta, </span><span>they can be set by adding the following in the section below </span><span class="c13 c12">[command]:</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span class="c13 c12">target_files = [</span></p><p class="c6"><span class="c13 c12">&nbsp; &nbsp; &quot;target1.fasta&quot;,</span></p><p class="c6"><span class="c13 c12">&nbsp; &nbsp; &quot;target2.fasta&quot;,</span></p><p class="c6"><span class="c13 c12">&nbsp; &nbsp; &quot;target3.fasta&quot;</span></p><p class="c6"><span class="c13 c12">]</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>Then, if your small RNA are in </span><span class="c12">smallRNA.fastq </span><span class="c0">you can run the following:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ hlsmallrna targetid smallRNA.fastq</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>Make sure the sequences in both </span><span class="c12">smallRNA.fastq</span><span class="c0">&nbsp;and each of the reference files have unique sequence ids, otherwise the command will fail.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>If you are working with very short sequences (&lt;5 bases), you will need to reduce the minimum length with the </span><span class="c12">-m </span><span class="c0">option. Conversely, if you know your minimum length is longer than this, this step can be sped up by increasing it. For example, if you know the shortest sequence you are dealing with is 10 bases long, you could run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ hlsmallrna targetid -m 10 smallRNA.fastq</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">Though this is not necessary and the program will produce a correct result for sequences longer than 5 without it, so only add if you need the extra speed.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>This produces </span><span class="c12">rna_target_list.csv </span><span class="c0">which contains a list of small RNA target pairs and some general information about each pair&rsquo;s alignment.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>If you are dealing with a small RNA that doesn&rsquo;t need perfect complementarity, you can add the </span><span class="c12">--num-mismatches</span><span class="c0">&nbsp;option to set the number of mismatches the alignment should allow. For example, if you want to allow up to four mismatches, you could run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c12">$ hlsmallrna targetid --num-mismatches 4 smallRNA.fastq</span></p><h2 class="c19" id="h.e77dx6e6zu8e"><span class="c3">All</span></h2><p class="c6"><span>Since the first three steps are often run one after the other, a shorthand </span><span>command </span><span class="c12">all </span><span>is</span><span class="c0">&nbsp;provided for convenience. This takes all of the arguments from the Process, Sort and Unitas steps, then runs them, piping the result from one into the next. All files produced by those steps are then produced in the output directory. </span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>For example if you have small RNAs in </span><span class="c12">smallRNA.fastq, </span><span>a genome in </span><span class="c12">genome.fasta</span><span>&nbsp;and </span><span class="c12">refseq </span><span>set in the config file and you want to trim the adapter sequence </span><span class="c12">ACATA </span><span class="c0">from both ends, but not run quality filtering you could run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ hlsmallrna all -c 0 -b ACATA smallRNA.fastq genome.fasta</span></p><p class="c6 c7"><span class="c13 c12"></span></p><h2 class="c19" id="h.g9ms4p1r7z4p"><span class="c3">Plotting Genome Location</span></h2><h3 class="c1" id="h.v8b0nssch3a2"><span>Basic Plot</span></h3><p class="c6"><span>A common task you might want to do is plotting the location of things within the genome, for example, genome features you are interested in or the positions of small RNAs you have aligned to the genome. To make this easy, a script is included in this environment called </span><span class="c12">build_coord_files</span><span>. This automatically generates tsv files for use with the R package </span><span class="c12">chromPlot</span><span class="c0">. This section explains how you can generate a plot with it.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>Firstly, you need at least two files: a full genome in a FASTA or FASTQ file and a set of locations you are interested in, either as an alignment output (in SAM or BAM format) or as feature annotations in GFF format. To start produce the genome coordinate file with the command: (assuming </span><span class="c12">genome.fasta</span><span class="c0">&nbsp;is your genome file)</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c12">$ build_coord_files -o genome_coord.tsv genome.fasta</span><span class="c0">&nbsp;</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>By default, the script will figure out what type of output to produce by the file extension. But if it can&rsquo;t, the user will need to provide a flag to tell it what type of input you are using (in this case </span><span class="c12">--fasta</span><span>). Next we produce the coordinates file for the item of interest, assuming it is in </span><span class="c12">alignment.sam </span><span class="c0">you can run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ build_coord_files -o alignment_coord.tsv alignment.sam</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>This will produce two files, one for coordinates on the sense strand and one for coordinates on the antisense strand. Finally, we can use these files to plot the result. If it isn&rsquo;t already, install </span><span class="c12">chromPlot </span><span class="c0">with bioconducter using the following R command:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">&gt; BiocManager::install(&quot;chromPlot&quot;)</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span class="c0">Then you should be able to create a plot using the following R code (of two example chromosomes called &lsquo;SRAE_chr1&rsquo; and &lsquo;SRAE_chr2&rsquo;):</span></p><p class="c6 c7"><span class="c34"></span></p><p class="c6"><span class="c2">chrom_coords &lt;- read.csv(&quot;/path/to/genome_coord.tsv&quot;, sep = &quot;\t&quot;)</span></p><p class="c6"><span class="c2">sense_pos &lt;- read.csv(&quot;/path/to/sense_alignment_coord.tsv&quot;, sep = &quot;\t&quot;)</span></p><p class="c6"><span class="c2">antisense_pos &lt;- read.csv(&quot;/path/to/antisense_alignment_coord.tsv&quot;, sep = &quot;\t&quot;)</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">library(chromPlot)</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">chromPlot(</span></p><p class="c6"><span class="c2">&nbsp; gaps = chrom_coords, annot1 = sense_pos, annot2 = antisense_pos, plotRndchr = T,</span></p><p class="c6"><span class="c2">&nbsp; chr = c(&quot;SRAE_chr1&quot;, &quot;SRAE_chr2&quot;), bin = 50000, </span></p><p class="c6"><span class="c2">&nbsp; chrSide = c(-1, 1, -1, -1, 1, -1, -1, 1)</span></p><p class="c6"><span class="c2">)</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">The script breaks down as follows:<br></span></p><ul class="c11 lst-kix_ddct7fkq5etb-0 start"><li class="c6 c14 li-bullet-0"><span>First three lines load in the data we just produced with </span><span class="c13 c12">build_coord_files</span></li></ul><p class="c6 c7 c29"><span class="c13 c12"></span></p><ul class="c11 lst-kix_ddct7fkq5etb-0"><li class="c6 c14 li-bullet-0"><span>Following line loads the </span><span class="c12">chromPlot</span><span class="c0">&nbsp;library into R</span></li></ul><p class="c6 c7 c29"><span class="c0"></span></p><ul class="c11 lst-kix_ddct7fkq5etb-0"><li class="c6 c14 li-bullet-0"><span>The final line plots the data with </span><span class="c13 c12">chromPlot:</span></li></ul><ul class="c11 lst-kix_ddct7fkq5etb-1 start"><li class="c6 c16 li-bullet-0"><span class="c12">gaps </span><span class="c0">defines what each chromosome should look like</span></li><li class="c6 c16 li-bullet-0"><span class="c12">annot1 </span><span>and </span><span class="c12">annot2</span><span class="c0">&nbsp;define the annotation data to show</span></li><li class="c6 c16 li-bullet-0"><span class="c12">plotRndchr</span><span>&nbsp;disables the assumption in &nbsp;</span><span class="c12">chromPlot </span><span class="c0">that the chromosomes are labelled with numbers</span></li><li class="c6 c16 li-bullet-0"><span class="c12">chr </span><span class="c0">selects which chromosomes to plot by specifying an array of names</span></li><li class="c6 c16 li-bullet-0"><span class="c12">bin </span><span class="c0">sets the bin size to group the features into when plotting</span></li><li class="c6 c16 li-bullet-0"><span class="c12">chrSide </span><span>defines which side of the chromosome to plot each feature, the only change we have made from the default is setting the second item to 1 as opposed to -1. This ensures </span><span class="c12">annot2</span><span>&nbsp;is plotted on the opposite side to </span><span class="c13 c12">annot1</span></li></ul><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span class="c0">The resulting plot will look something like:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 603.21px; height: 461.33px;"><img alt="" src="images/image1.png" style="width: 603.21px; height: 461.33px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><h3 class="c1" id="h.3ax5siuj26a8"><span class="c18">Dealing with Multiple Scaffolds</span></h3><p class="c6"><span>In more complex cases, one or more of your chromosomes may be split into multiple scaffolds. </span><span class="c12">Build_coord_files </span><span>has a scaffold aware mode that can automatically try to merge these, but it will only work if your scaffold&rsquo;s IDs are in the form of </span><span class="c12">&lt;chromosome_name&gt;_scaffold&lt;scaffold_number&gt;</span><span>&nbsp;e.g. </span><span class="c12">SRAE_chrX_scaffold1</span><span>. Once you have this, you can generate the data files with the </span><span class="c12">-c </span><span class="c0">flag set, as follows:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c12">$ build_coord_files -c -o genome_coord.tsv genome.fasta</span><span class="c0">&nbsp;</span></p><p class="c6"><span class="c13 c12">$ build_coord_files -c --genome-coords genome_coord.tsv -o alignment_coord.tsv alignment.sam</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>Note you need to provide the genome coords file when generating files for what you are plotting, this allows for a correction of coordinates to be made from where scaffolds are merged. In addition to the regular files, this produces </span><span class="c12">scaffold_info_genome_coord.tsv</span><span class="c0">, that can be used to add scaffold dividers to your plots. To do this, the following R script can be run:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c2">chrom_coords &lt;- read.csv(&quot;/path/to/genome_coord.tsv&quot;, sep = &quot;\t&quot;)</span></p><p class="c6"><span class="c2">sense_pos &lt;- read.csv(&quot;/path/to/sense_alignment_coord.tsv&quot;, sep = &quot;\t&quot;)</span></p><p class="c6"><span class="c2">antisense_pos &lt;- read.csv(&quot;/path/to/antisense_alignment_coord.tsv&quot;, sep = &quot;\t&quot;)</span></p><p class="c6"><span class="c2">scaffolds &lt;- read.csv(&quot;/path/to/scaffold_info_genome_coord.tsv&quot;, sep = &quot;\t&quot;)</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">library(chromPlot)</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">chromPlot(</span></p><p class="c6"><span class="c2">&nbsp; gaps = chrom_coords, annot1 = sense_pos, annot2 = antisense_pos, plotRndchr = T,</span></p><p class="c6"><span class="c2">&nbsp; chr = c(&quot;SRAE_chr1&quot;, &quot;SRAE_chr2&quot;), bin = 50000, bands = scaffolds,</span></p><p class="c6"><span class="c2">&nbsp; chrSide = c(-1, 1, -1, -1, 1, -1, -1, 1)</span></p><p class="c6"><span class="c12 c43">)</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">Compared to the last script, this adds:</span></p><ul class="c11 lst-kix_m6ci32kd50uv-0 start"><li class="c6 c14 li-bullet-0"><span class="c0">The fourth line, that loads in the scaffold data</span></li></ul><p class="c6 c7 c29"><span class="c0"></span></p><ul class="c11 lst-kix_m6ci32kd50uv-0"><li class="c6 c14 li-bullet-0"><span>The </span><span class="c12">bands </span><span>argument, that draws red lines at the end of each scaffold</span></li></ul><p class="c6 c7 c29"><span class="c0"></span></p><p class="c6"><span>Note that, by default the bands are 20,000 bp wide, as that looks good on the plot. At that size they may not appear if you make the bins too much bigger. Therefore, if you plan to make the bins bigger, pass a larger value to the </span><span class="c12">--band-width</span><span class="c0">&nbsp;flag when creating the genome file. That makes the bands bigger, which results in a more visible line on the plot.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">A plot produced by this method will look something like:</span></p><p class="c6"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 308.00px; height: 335.95px;"><img alt="" src="images/image2.png" style="width: 603.96px; height: 462.95px; margin-left: 0.00px; margin-top: -56.95px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><hr style="page-break-before:always;display:none;"><p class="c24 c7 title" id="h.xxke3cbl51hz"><span class="c28"></span></p><h1 class="c10" id="h.m9e2tyi06fss"><span class="c33">Further Processing Options</span></h1><h2 class="c19" id="h.mu0rcgmaopdr"><span>Signature Identification</span></h2><h3 class="c1" id="h.j40p9cyavwlb"><span>AT/GC Richness</span><span class="c23">&nbsp;</span></h3><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c36">Nucleotide richness and the presence of any conserved motifs is identified using WebLogo (</span><span class="c42"><a class="c5" href="https://www.google.com/url?q=http://weblogo.threeplusone.com/&amp;sa=D&amp;source=editors&amp;ust=1664380231734256&amp;usg=AOvVaw3fO5GHR30o2b0sX2-O6tgw">weblogo.threeplusone.com</a></span><span class="c27">). Sequences need to be the same length and can be either uploaded using a fasta file or can be pasted in. </span></p><h3 class="c1" id="h.4r7fhoddcf9x"><span class="c18">Upstream motif (e.g. piRNA)</span></h3><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">map to genome with bowtie2, convert sam to bam with samtools, convert bamToBed with bedtools </span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ bowtie2 -x genome.fasta -f -U sequences.fasta -S sequences_mapped.sam</span></p><p class="c6"><span class="c13 c12">$ samtools view sequences_mapped.sam -o sequences_mapped.bam</span></p><p class="c6"><span class="c13 c12">$ bamToBed -i sequences.bam &gt; sequences.bed</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">extract flanking sequences with flankBed in bedtools (need to use bedtools format genome file - generate with faidx and cut)</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ samtools faidx genome.fasta &nbsp;</span></p><p class="c6"><span class="c13 c12">$ cut -f 1,2 genome.fa.fai &gt; chrom.sizes</span></p><p class="c6"><span class="c13 c12">$ flankBed -i sequences.bed -g chrom.sizes -b 100 &gt; sequences_100nt_flank.bed</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">or 60 upstream only</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ flankBed -i sequences.bed -g chrom.sizes -l 60 -r 0 &gt; sequnences_60nt.bed</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">extract fasta sequences from bed fastFromBed with bedtools</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ fastaFromBed -fi genome.fasta -bed sequences_100nt_Flank.bed -fo sequences_100nt_Flank_seqs.fasta</span></p><h3 class="c1" id="h.uqa0lrhum6ba"><span>Ping Pong signature</span></h3><p class="c6"><span class="c40">To find ping-pong signature, Unitas with the option -pp can be used to find </span><span class="c0">5&rsquo; overlaps of mapped sequence reads to the genome and calculate a Z-score for the enrichment of 10 bp overlaps.</span></p><p class="c38"><span class="c13 c12">$ perl unitas.pl -pp -input map.file (SAM or ELAND3 format) &ndash;species x &ndash;refseq reference.fasta</span></p><h3 class="c1" id="h.k4lth3h5hxsf"><span>Dicer Signature (StepRNA)</span></h3><p class="c6"><span>To install stepRNA, see the documentation on the GitHub page (</span><span class="c26"><a class="c5" href="https://www.google.com/url?q=https://github.com/bmm514/stepRNA%23Installation&amp;sa=D&amp;source=editors&amp;ust=1664380231738327&amp;usg=AOvVaw1RAzllLmkxZsgBHJmR9pjw">https://github.com/bmm514/stepRNA</a></span><span class="c0">).</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">stepRNA is able to identify a Dicer processing signature from a small RNA sequence dataset. It requires two FASTA files as input that have had their adapters already trimmed from the ends.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ stepRNA -r/--reference REFERENCE.fa -q/--reads READS.fa</span></p><p class="c6 c7"><span class="c13 c12"></span></p><ul class="c11 lst-kix_qxq34zax48bo-0 start"><li class="c6 c14 li-bullet-0"><span class="c12">READS.fa</span><span class="c0">&nbsp;should contain all of the small RNA sequencing reads.</span></li><li class="c6 c14 li-bullet-0"><span class="c12">REFERENCE.fa</span><span class="c0">&nbsp;is usually filtered to investigate small RNAs of interest e.g. 26G sRNAs.</span></li></ul><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>Identical reads in the READ and REFERENCE files can also be removed before searching for a Dicer signature, if desired, by using </span><span class="c12">-e/--remove_exact</span><span class="c0">. </span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c15">Importantly the FASTA headers must be unique</span><span>&nbsp;- this can be done by stepRNA with </span><span class="c12">-u/--make_unique</span><span class="c0">.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">stepRNA will then generate: </span></p><ul class="c11 lst-kix_pbqqzrwh3zqo-0 start"><li class="c6 c14 li-bullet-0"><span class="c0">BAM alignment files (with different combinations of overhang lengths)</span></li><li class="c6 c14 li-bullet-0"><span class="c0">Overhang length CSV</span></li><li class="c6 c14 li-bullet-0"><span class="c0">Passenger Read Length CSV</span></li></ul><p class="c6"><span class="c0">These can then be used to plot the distribution and/or further explore the reads that have aligned. </span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c15">Note: </span><span>If the sRNA lengths in the read file are all the same i.e. 21 nt long, stepRNA will not be able to find overhangs due to the scoring system. In this case the </span><span class="c12">-m/--min_score </span><span class="c0">should be set to an appropriate value; we recommend sRNA length - 7nt (e.g. 21 - 6 = 15).</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>$ </span><span class="c12">stepRNA --help </span><span class="c0">to bring up the command line help, a description of the methods can be found in the stepRNA publication (add_when_available)</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c2">stepRNA [-h] -r REFERENCE -q READS [-n NAME] [-d DIRECTORY]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;[-m MIN_SCORE] [-e] [-u] [-j] [-V]</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">Align a reference RNA file to read sequences. Output will be a set of CSV files containing information about the length of the reads, number of reads aligned to a reference sequence and the length of overhangs of the alignment.</span></p><p class="c6"><span class="c2">Reference RNA file will be automatically indexed</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">Optional Arguments:</span></p><p class="c6"><span class="c2">&nbsp; -h, --help &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;show this help message and exit</span></p><p class="c6"><span class="c2">&nbsp; -n NAME, --name NAME &nbsp;Prefix for the output files</span></p><p class="c6"><span class="c2">&nbsp; -d DIRECTORY, --directory DIRECTORY</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Directory to store the output files</span></p><p class="c6"><span class="c2">&nbsp; -m MIN_SCORE, --min_score MIN_SCORE</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Minimum score to accept, default is the shortest read</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; length</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">Required Arguments:</span></p><p class="c6"><span class="c2">&nbsp; -r REFERENCE, --reference REFERENCE</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Path to the reference sequences</span></p><p class="c6"><span class="c2">&nbsp; -q READS, --reads READS</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Path to the read sequences</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">Flags:</span></p><p class="c6"><span class="c2">&nbsp; -e, --remove_exact &nbsp; &nbsp;Remove exact read matches to the reference sequence</span></p><p class="c6"><span class="c2">&nbsp; -u, --make_unique &nbsp; &nbsp; Make FASTA headers unique in reference and reads i.e.</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &gt;Read_1 &gt;Read_2</span></p><p class="c6"><span class="c2">&nbsp; -j, --write_json &nbsp; &nbsp; &nbsp;Write count dictionaries to a JSON file</span></p><p class="c6"><span class="c2">&nbsp; -V, --version &nbsp; &nbsp; &nbsp; &nbsp; Print version number then exit.</span></p><p class="c6 c7"><span class="c0"></span></p><h3 class="c1" id="h.wcf5w2j3zywp"><span class="c18">Same Strand Overlap</span></h3><p class="c6"><span>If you wish to </span><span>detect same</span><span>&nbsp;strand overlaps, you can use the </span><span class="c12">overlap_ss</span><span class="c0">&nbsp;script that is installed in the same environment as this pipeline. This uses a number of bash programs to create files that show where two sets of small RNA that appear on the same strand overlap.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">To use this script you need two sets of small RNA as FASTA or FASTQ files, one for the base and one for the overlap, plus the genome of the species of interest. You can then run the script in the following way:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">$ overlap_ss genome.fasta smallRNA1.fastq smallRNA2.fastq</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>This script will then create output in the output directory specified in </span><span class="c12">config.toml </span><span>under the directory </span><span class="c12">samestrand_overlap/.</span></p><h3 class="c1" id="h.po793ufjxkq3"><span class="c18">miRNA</span></h3><h1 class="c10" id="h.ckvkzxv8xboy"><span class="c33">Notes on Settings</span></h1><h2 class="c19" id="h.dqg4gw5x0t24"><span class="c3">Configuration file</span></h2><p class="c6"><span>For ease of use this program requires a configuration file in TOML format to run successfully. At minimum it can be a blank file, but I strongly suggest setting at least the </span><span class="c12">output_directory</span><span>&nbsp;setting to be named something more useful than </span><span class="c12">output,</span><span class="c0">&nbsp;the default setting.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>By default the program tries to use </span><span class="c12">config.toml</span><span>&nbsp;in your current directory, but it can be set to a different file using the </span><span class="c12">-C </span><span class="c0">option before specifying which task to run, e.g.:</span></p><p class="c6 c7"><span class="c12 c13"></span></p><p class="c6"><span class="c13 c12">$ hlsmallrna -C alternate.toml sort ...</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span>The layout of a TOML file has a number of sections each with a number of configuration keys. Each key has an expected type (e.g. string, integer, floating point number etc.) and the program will check they are as expected before running. If one is wrong, the program will crash and alert the user of the problem. A list of these keys and example values can be found in the </span><span class="c12">README.md</span><span class="c0">&nbsp;included with the program.</span></p><p class="c6 c7"><span class="c0"></span></p><h2 class="c19" id="h.8xpup46t5df5"><span class="c3">Manually setting parameters of internal commands</span></h2><p class="c6"><span>For advanced uses, you may want to add parameters to the internal commands run. This can be done through variables in the </span><span class="c12">cli-tools</span><span>&nbsp;section of the configuration file. A full list of these parameters can be found in the </span><span class="c12">README.md</span><span class="c0">&nbsp;provided with the code. </span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>There are two main ones that you can set </span><span class="c12">&lt;name&gt;_pass_threads</span><span>&nbsp;and </span><span class="c12">&lt;name&gt;_params</span><span class="c0">. The former is a simple boolean that can be set to false if you want the default number of threads for the tool, instead of the number set in the general section of the configuration file. The latter is a list of custom parameters to be added on to the command when called. For example if you wanted to set min-length to 10 and disable threads for FastQC you could add the following to the configuration file:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c13 c12">[cli-tools]</span></p><p class="c6"><span class="c13 c12">[cli-tools.fastqc]</span></p><p class="c6"><span class="c13 c12">fastqc_pass_threads = false</span></p><p class="c6"><span class="c12">fastqc_params = [&quot;--min_length&quot;, &quot;10&quot;]</span></p><h2 class="c19" id="h.u09whqqmb0i9"><span class="c3">Suppressing excess output</span></h2><p class="c6"><span>If you do not want to view the output from each of the commands, you can pass the </span><span class="c12">-q </span><span>option to the main script. (Before the task selection, like with </span><span class="c12">-C</span><span class="c0">). It is recommended to test the command first however, as this option sometimes suppresses useful error messages.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>Note this doesn&rsquo;t stop all output, only that of internal commands. If you don&rsquo;t want any progress messages whatsoever, only errors and warnings from python, set </span><span class="c12">-q </span><span>twice. E.g. </span><span class="c12">-q -q </span><span>or </span><span class="c13 c12">-qq.</span></p><p class="c6 c7"><span class="c0"></span></p><h2 class="c19" id="h.c34jicjkp0we"><span class="c3">Troubleshooting</span></h2><p class="c6"><span class="c0">If it is taking too long to run or producing odd data, I strongly suggest deleting or moving the output directory somewhere else, then doing a clean run. This is because some steps use the data in the output directory as input, so may be combining old and new data.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">When using the targetid command, samtools is used to filter out duplicates and create the FASTQ file outputs. Samtools is known to be picky about how the input file is layed out, so will fail. The program should keep running, but if you want these files, check the following in </span></p><p class="c6"><span class="c0">your input files:<br></span></p><ul class="c11 lst-kix_6thhdslfbwlj-0 start"><li class="c6 c14 li-bullet-0"><span class="c0">For FASTA files, there is no @ at the start of any of the header strings</span></li><li class="c6 c14 li-bullet-0"><span class="c0">There are no duplicate headers in any of the input files</span></li></ul><hr style="page-break-before:always;display:none;"><h1 class="c10" id="h.yn1evvqauc8s"><span class="c33">CLI Reference</span></h1><p class="c6"><span class="c0">References of how to use the command line interface of the program and notes on some ways particular commands work that could trip you up.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">General help message is as follows:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c2">usage: hlsmallrna [-h] [-q] [-C PATH_TO_CONFIG]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;{process,sort,extractnc,unitas,targetid,all} ...</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">Pipeline to process small RNAs</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">positional arguments:</span></p><p class="c6"><span class="c2">&nbsp; {process,sort,extractnc,unitas,targetid,all}</span></p><p class="c6"><span class="c2">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;process &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Preprocessing for the RNA</span></p><p class="c6"><span class="c2">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;sort &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Find RNAs that align to a genome and sort them by</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;length</span></p><p class="c6"><span class="c2">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;extractnc &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Extract the noncoding region from a fasta with a GFF</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;file</span></p><p class="c6"><span class="c2">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;unitas &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Run unitas on split files and merge results</span></p><p class="c6"><span class="c2">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;targetid &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Align small RNA to a number of genome features to find</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;out what is targeted</span></p><p class="c6"><span class="c2">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;all &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Run process, sort and unitas one after the other</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">optional arguments:</span></p><p class="c6"><span class="c2">&nbsp; -h, --help &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;show this help message and exit</span></p><p class="c6"><span class="c2">&nbsp; -q, --quiet &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Suppress output from intermediate commands</span></p><p class="c6"><span class="c2">&nbsp; -C PATH_TO_CONFIG, --path-to-config PATH_TO_CONFIG</span></p><p class="c6"><span class="c12 c43">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Path to the TOML format config file to use</span></p><h2 class="c19" id="h.rt1yuq1t17b5"><span class="c3">Process</span></h2><p class="c6"><span class="c0">Docs:</span></p><p class="c6"><span class="c2">process [-h] [-a ADAPTER] [-g FRONT] [-b ANYWHERE]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[-c CUTOFF]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;small_rna</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">positional arguments:</span></p><p class="c6"><span class="c2">&nbsp; small_rna &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Path to FASTQ containing the small RNA</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">optional arguments:</span></p><p class="c6"><span class="c2">&nbsp; -h, --help &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;show this help message and exit</span></p><p class="c6"><span class="c2">&nbsp; -a ADAPTER, --adapter ADAPTER</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sequence of the adapter to remove from the 3&#39; end</span></p><p class="c6"><span class="c2">&nbsp; -g FRONT, --front FRONT</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sequence of the adapter to remove from the 5&#39; end</span></p><p class="c6"><span class="c2">&nbsp; -b ANYWHERE, --anywhere ANYWHERE</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sequence of the adapters to remove from both ends</span></p><p class="c6"><span class="c2">&nbsp; -c CUTOFF, --cutoff CUTOFF</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Quality cutoff to trim RNA sequences at</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">Input files:</span></p><p class="c6 c35"><span class="c12">Small_rna </span><span class="c0">- &nbsp;fastq file containing raw RNA-seq data</span></p><p class="c6 c7 c35"><span class="c0"></span></p><p class="c6"><span class="c0">Output files:</span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Fastqc/ -</span><span class="c0">&nbsp;directory containing the raw output of FastQC</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Trimmed_rna.fastq - </span><span class="c0">file containing the raw RNA with adapters removed (only produced if an adapter sequence is provided)</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Cut_sequences.fastq - </span><span class="c0">sequences with low quality parts removed</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>Note that the parts of this step can easily be skipped. If you want to skip adapter trimming, don&rsquo;t specify any adapters with </span><span class="c12">-a, -g </span><span>or </span><span class="c12">-c. </span><span>If you don&rsquo;t want to run fastQC set </span><span class="c12">-c 0</span><span class="c0">. The program knows that in these cases, the steps do not have to be run.</span></p><h2 class="c19" id="h.v1awwydjuyb5"><span class="c3">Sort</span></h2><p class="c6"><span class="c2">sort [-h] [-l MIN_LENGTH] [-x MAX_LENGTH] small_rna genome</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">positional arguments:</span></p><p class="c6"><span class="c2">&nbsp; small_rna &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Path to FASTQ containing the small RNA</span></p><p class="c6"><span class="c2">&nbsp; genome &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Genome to align against</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">optional arguments:</span></p><p class="c6"><span class="c2">&nbsp; -h, --help &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;show this help message and exit</span></p><p class="c6"><span class="c2">&nbsp; -l MIN_LENGTH, --min-length MIN_LENGTH</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Minimum length to bin</span></p><p class="c6"><span class="c2">&nbsp; -x MAX_LENGTH, --max-length MAX_LENGTH</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Maximum length to bin</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c0">Input files:</span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Small_rna - </span><span class="c0">fastq file containing RNA with adapters removed</span></p><p class="c6"><span class="c0">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Genome - </span><span class="c0">Reference genome of the species you are using to align against in fasta format</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">Output files:</span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Bbmap_index/ - </span><span class="c0">contains the index of the reference genome created by bbmap</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Mapped_sequences.fastq - </span><span class="c0">sequences successfully mapped to the reference by bbmap</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">binned_rna/ - </span><span class="c0">directory containing one fastq file for each length of sequence contained in the bbmap output</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12 c22">Rna_length_report.csv</span><span class="c17">&nbsp;- </span><span class="c8">table showing a summary of the RNAs by length and first base</span></p><p class="c6 c7"><span class="c8"></span></p><p class="c6"><span class="c22">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12 c22">Baseplot.png - </span><span class="c8">plot of the length and first base of the RNAs, using the data in rna_length_report.csv</span></p><p class="c6 c7"><span class="c8"></span></p><p class="c6"><span class="c22">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12 c22">Baseplot_data.csv - </span><span class="c8">raw data used to make baseplot.png to allow for easy regraphing</span></p><p class="c6 c7"><span class="c8"></span></p><h2 class="c19" id="h.4iigvduhoact"><span class="c3">ExtractNC</span></h2><p class="c6"><span class="c0">Docs:</span></p><p class="c6"><span class="c2">extarctnc [-h] genome gff_file</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">positional arguments:</span></p><p class="c6"><span class="c2">&nbsp; genome &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;FASTA containing the genome to extract from</span></p><p class="c6"><span class="c2">&nbsp; gff_file&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GFF file containing annotations of CDS and mRNA regions</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">optional arguments:</span></p><p class="c6"><span class="c2">&nbsp; -h, --help &nbsp;show this help message and exit</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c0">Input files:</span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Genome - </span><span class="c0">file containing the whole genome of the species of interest</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c12">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gff_file - </span><span class="c0">GFF3 file containing annotations at least for the mRNA and coding region (labelled CDS) for the genome</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">Output files:</span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Noncoding.fasta - </span><span class="c0">FASTA file containing only the transcribed noncoding regions of the DNA</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">Note the method used here will not work if there a coding region is labelled outside a single mRNA region. The program does a check before running to make sure this is true and throws an exception before trying to run if so, to prevent nonsense results from being produced.</span></p><h2 class="c19" id="h.9bjdtm5xm2gh"><span class="c3">Unitas</span></h2><p class="c6"><span class="c0">Docs:</span></p><p class="c6"><span class="c2">unitas [-h] [-d CDS] [-u UNSPLICED_TRANSCRIPTOME]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[-r [REFSEQ [REFSEQ ...]]] [-s SPECIES]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;path_to_rnas</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">positional arguments:</span></p><p class="c6"><span class="c2">&nbsp; path_to_rnas &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Path to the folder with varying length RNAs in</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">optional arguments:</span></p><p class="c6"><span class="c2">&nbsp; -h, --help &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;show this help message and exit</span></p><p class="c6"><span class="c2">&nbsp; -d CDS, --cds CDS &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Optional CDS region, passed to unitas</span></p><p class="c6"><span class="c2">&nbsp; -u UNSPLICED_TRANSCRIPTOME, --unspliced-transcriptome UNSPLICED_TRANSCRIPTOME</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Optional, unspliced transcriptome, passed to unitas</span></p><p class="c6"><span class="c2">&nbsp; -r [REFSEQ [REFSEQ ...]], --refseq [REFSEQ [REFSEQ ...]]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;References for use with unitas</span></p><p class="c6"><span class="c2">&nbsp; -s SPECIES, --species SPECIES</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Species to set in unitas arguments</span></p><p class="c6"><span class="c0">Input files:</span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Path_to_rnas - </span><span class="c0">path to the directory containing fastq files of each set of &nbsp;small RNAs you want to procesus in unitas</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Note: to get unitas to work, you need to set either the species to a compatible species or provide one or more files of reference small RNAs. These can either be done with the </span><span class="c12">-s </span><span>and </span><span class="c12">-r </span><span class="c0">optional arguments or by adding them to your configuration file (see below).</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c12">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;CDS - </span><span class="c0">Coding sequence, pipeline automatically labels with Gene and combines with the unspliced transcriptome, optional</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Unspliced_transcriptome - </span><span>Unspliced transcriptome, pipeline automatically labels with Gene and combines with the CDS, optional</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">Output files:</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">unitas/ - </span><span class="c0">directory containing the results for all the unitas runs performed by the program</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Unitas_summery.csv - </span><span class="c0">CSV file containing the combined summaries of all of the unitas runs (plus some extra numbers). It is recommended to open it in a spreadsheet package (e.g. microsoft excel, libreoffice calc) as there is a lot of data and it may look a mess in a text editor.</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">UnitasGraph.png - </span><span class="c0">graph showing the categories allocated by unitas against the length of the small RNA</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Unitas_graph_data.csv - </span><span class="c0">raw data used for creating unitasGraph.png to allow for easy regraphing</span></p><h2 class="c19" id="h.h79krq96jgla"><span class="c3">TargetID</span></h2><p class="c6"><span class="c0">Docs:</span></p><p class="c6"><span class="c2">targetid [-h] [-m MIN_SEQ_LENGTH] [-t TARGET_FILES [TARGET_FILES ...]]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[--num-mismatches NUM_MISMATCHES]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;small_rna</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">positional arguments:</span></p><p class="c6"><span class="c2">&nbsp; small_rna &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Path to the FASTQ containing the small RNA to find targets of</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">optional arguments:</span></p><p class="c6"><span class="c2">&nbsp; -h, --help &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;show this help message and exit</span></p><p class="c6"><span class="c2">&nbsp; -m MIN_SEQ_LENGTH, --min-seq-length MIN_SEQ_LENGTH</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Minimum sequence length to properly align</span></p><p class="c6"><span class="c2">&nbsp; -t TARGET_FILES [TARGET_FILES ...], --target-files TARGET_FILES [TARGET_FILES ...]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Files containing genome features that could be targeted</span></p><p class="c6"><span class="c2">&nbsp; --num-mismatches NUM_MISMATCHES</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Number of mismatches to allow in the alignment, defaults to 0</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c0">Input: </span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">small_rna</span><span class="c0">&nbsp;- fastq file containing small RNA to look for targets of</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">target_files</span><span class="c0">&nbsp;- one or more fastq files containing a list of potential target sequences</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span class="c0">Output:</span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">Bowtie_indexes/</span><span class="c0">&nbsp;- indexes produced of the target files with bowtie2</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">target_alignments/</span><span class="c0">&nbsp;- SAM files containing the results of alignment attempts and FASTQ files containing all the small RNA successfully aligned against the targets. One set of files are produced for each target file provided</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span><span class="c12">rna_target_list.csv</span><span class="c0">&nbsp;- list of RNA target pairs</span></p><p class="c6 c7"><span class="c0"></span></p><p class="c6"><span>Note that the </span><span class="c12">samtools view </span><span>command used in this step is the only one that has defaults pre-set in the </span><span class="c12">samtools_view_params </span><span>config key, namly </span><span class="c12">-h -F 256 -F 4</span><span>. They will be removed if you set the value differently in the config file. To only add parameters, you will need to set these ones again before the extra ones you want to set. E.g. </span><span class="c13 c12">samtools_view_params = [&quot;-h&quot;, &quot;-F&quot;, &quot;256&quot;, &quot;-F&quot;, &quot;4&quot;, ...]</span></p><p class="c6 c7"><span class="c13 c12"></span></p><p class="c6"><span class="c0">Make sure all of your target files are labelled with what they are, by prepending the name and | (the bar or pipe character) to the ID in the fasta file. If you don&rsquo;t do this, it will not be nicely sorted in the results correctly.</span></p><p class="c6 c7"><span class="c0"></span></p><h2 class="c19" id="h.jyq65jwuj5bd"><span class="c3">Label for Unitas</span></h2><p class="c6"><span class="c2">usage: label_for_unitas [-h] [-o OUTPUT] label file_path</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">Program to automatically prepend unitas labels to all the sequences in a FASTA</span></p><p class="c6"><span class="c2">file</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">positional arguments:</span></p><p class="c6"><span class="c2">&nbsp; label &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Label to prepend to the sequences in a fasta file</span></p><p class="c6"><span class="c2">&nbsp; file_path &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Path to the fasta file to label</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">optional arguments:</span></p><p class="c6"><span class="c2">&nbsp; -h, --help &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;show this help message and exit</span></p><p class="c6"><span class="c2">&nbsp; -o OUTPUT, --output OUTPUT</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;File to output to, defaults to labelled.fasta</span></p><p class="c6 c7"><span class="c2"></span></p><h2 class="c19" id="h.h5pmlelo7os6"><span class="c3">Build Coord Files</span></h2><p class="c6"><span class="c2">usage: build_coord_files [-h] [-a] [-q] [-s] [-b] [-g] [-r] [-c]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[--feature FEATURE] [--genome-coords GENOME_COORDS]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[--band-width BAND_WIDTH] [-o OUTPUT]</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;input_file</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">Convert common bioinformatics file formats into coordinate file to plot</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">positional arguments:</span></p><p class="c6"><span class="c2">&nbsp; input_file &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;File to convert, attempts to autodetect type</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">optional arguments:</span></p><p class="c6"><span class="c2">&nbsp; -h, --help &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;show this help message and exit</span></p><p class="c6"><span class="c2">&nbsp; -a, --fasta &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Treat input as a FASTA file</span></p><p class="c6"><span class="c2">&nbsp; -q, --fastq &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Treat input as a FASTQ file</span></p><p class="c6"><span class="c2">&nbsp; -s, --sam &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Treat input as a SAM file</span></p><p class="c6"><span class="c2">&nbsp; -b, --bam &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Treat input as a BAM file</span></p><p class="c6"><span class="c2">&nbsp; -g, --gff &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Treat input as a GFF file</span></p><p class="c6"><span class="c2">&nbsp; -r, --rm-fa-out &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Treat input as a RepeatMasker .fa.out file</span></p><p class="c6"><span class="c2">&nbsp; -c, --scaffold-aware &nbsp;Merge scaffolds into one chromosome</span></p><p class="c6"><span class="c2">&nbsp; --feature FEATURE &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Select a gff feature to use</span></p><p class="c6"><span class="c2">&nbsp; --genome-coords GENOME_COORDS</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;File containing the coordinates of the genome</span></p><p class="c6"><span class="c2">&nbsp; --band-width BAND_WIDTH</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Size of the bands showing change in scaffolds</span></p><p class="c6"><span class="c2">&nbsp; -o OUTPUT, --output OUTPUT</span></p><p class="c6"><span class="c2">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;File to output to</span></p><h2 class="c19" id="h.mkenrz55dxfs"><span class="c3">Overlap SS</span></h2><p class="c6"><span class="c2">usage: overlap_ss [-h] [-q] genome rna_file_1 rna_file_2</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">Looks for overlaps of two classes of RNA on the same strand of DNA</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">positional arguments:</span></p><p class="c6"><span class="c2">&nbsp; genome &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;FASTA file containing the genome of the organism</span></p><p class="c6"><span class="c2">&nbsp; rna_file_1 &nbsp; File containing the RNA to use as a base</span></p><p class="c6"><span class="c2">&nbsp; rna_file_2 &nbsp; File containing the RNA to look for overlaps with</span></p><p class="c6 c7"><span class="c2"></span></p><p class="c6"><span class="c2">optional arguments:</span></p><p class="c6"><span class="c2">&nbsp; -h, --help &nbsp; show this help message and exit</span></p><p class="c6"><span class="c2">&nbsp; -q, --quiet &nbsp;Suppress output from intermediate commands</span></p><div><p class="c7 c41"><span class="c0"></span></p></div></body></html>